docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,116 @@
+
+### Association Changes
+
+Today, a device profile is associated to a device via its name (the ProfileName on the Device struct).  The profile can also associated to events and readings (via the ProfileName in the Event and Reading objects).
+
+Profiles can be associated to more than one device.  Profiles can be associated to multiple events/readings.
+
+Rules about change of a device profile association have never been explicitly defined.  While the section above outline what can/cannot be changed in the profile and under what circumstances, this section outlines whether a device, event or reading can alter the associated device profile and under what circumstances.
+
+Change of the device profile on a device – never allowed unless there are no events or readings in the system associated to the device profile or its device resources or commands.
+
+A provision watcher also has an association to a device profile.  However, the device profile is only used to know the type of device to create in case of provisioning.  Therefore, if a device profile changes, it does not impact the provision watcher.
+
+A device profile cannot be deleted while it is associated to a provision watcher
+
+Change of the device profile on an event or reading is never allowed.  Events and readings are immutable.
+
+### Applicability
+
+This document dictates the rules around allowed change to device profiles (and associations).  All EdgeX services will abide by these rules (implementing appropriate checks and validation as necessary to enforce the rules) when this ADR is approved to include:
+- service APIs
+- user interfaces (like the Edge GUI)
+- client tools (like the EdgeX CLI)
+
+## Proposed Design
+
+The following metadata API changes are suggested as a means to implement the device profile rules of this ADR.
+
+- Block Device DELETE when there is any associated event/reading exists.
+- Block Profile (Upload) PUT when there is any associated device/event/reading exists.
+- Block Profile DELETE when there is any associated device/event/reading exists.
+- Allow empty Profile POST (containing no device resources or commands)
+- Add Profile General Property PATCH API
+- Add Profile Device Resource POST API
+- Add Profile Device Resource PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Resource DELETE API (check it's not associated to any device, reading or event)
+- Add Profile Device Command POST API
+- Add Profile Device Command PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Command DELETE API (check it's not associated to any device, reading or event)
+
+## Decision
+
+TBD
+
+## Consequences/Considerations
+
+No validation on a device profile PUT will occur based on the terms above.  If you want to add a new device resource, you use the add device   resource API (POST).  You cannot add a device resource through device profile PUT calls unless there are no devices associated to the device profile.
```

Comment:
```suggestion
No change validation on a device profile PUT will occur based on the terms above.  If you want to add a new device resource, you use the add device   resource API (POST).  You cannot add a device resource through device profile PUT calls unless there are no devices associated to the device profile.
```

Reply:
Hmmm, I'd rather see us not allow device resources or commands to be added via `PUT` at all as we were told implementing the checks are hard, and we can't just trust the end-user that there are no devices associated...

Reply:
I added @lenny-intel suggested "change" word.  I listened to the recording several times.  We went back and forth on PUT vs PATCH.  @cloudxxx8 - can you weigh in with your preference in the comments here and why again here and we'll finalize in the Core WG meeting.

Reply:
@tonyespy , agreed. Maybe should be `detailed change validation` , i.e. it does what it currently does which is only allows if no associated device, provision watcher or event.

I didn't want `No validation` to be taken as it doesn't do the standard request data validation, which we are not chnaging.

Reply:
I would still prefer if we only had one way to modify device profiles. It seems crazy to have some API methods perform fine-grained validation, and other global APIs (eg. device profile PUT) only offer course-grained validation. I think we're overthinking this...

Reply:
@tonyespy for example, if the user just want to add a new Device Command to provide different Device Resource aggregation, they have to query the existing big profile (it might contain 1000 Device Resource) and edit the file to add one Device Command section, and call the PUT.  It's much more crazy.